src/controllers/midi/midicontroller.cpp

```diff
@@ -275,23 +300,53 @@ void MidiController::processInputMapping(const MidiInputMapping& mapping,
             return;
         }
 
-        QJSValue function = pEngine->wrapFunctionCode(mapping.control.item, 5);
-        const auto args = QJSValueList{
-                channel,
-                control,
-                value,
-                status,
-                mapping.control.group,
-        };
-        if (!pEngine->executeFunction(&function, args)) {
-            qCWarning(m_logBase) << "MidiController: Invalid script function"
-                                 << mapping.control.item;
-        }
+        return std::visit(
+                MidiUtils::overloaded{
+                        [this, pEngine, channel, control, value, status](
+                                ConfigKey configKey) {
```

Comment:
make clazy happy.
```suggestion
                                const ConfigKey& configKey) {
```

Reply:
I can't make it a `const`. `pEngine->executeFunction` expects a not const pointer. Is `ConfigKey& configKey` ok or there's another way?

Reply:
I don't see how thats a problem. you can construct the non-const `QJSValue function` from a `const configKey.item`, or a I wrong? Do you get a compiler error?

Reply:
Yup. [On that block](https://github.com/mixxxdj/mixxx/pull/12781/files/c2a31a0cafcab566dd21b4406a2fa689b923eadb#diff-ee87c49f05ff8aa897f732378f814f8e917cd4ec0b014b3021cda3c3e839dc7cR325-R340). I can have `[this, pEngine, channel, control, value, status](QJSValue& function) { ... }`, but not `const QJSValue& function`. Is this a problem?

Reply:
Ah right, there it makes sense. to take by mutable reference (`QJSValue&`). but the snippet we're talking about is the ConfigKey which doesn't need to be taken by mutable ref. I think they can be different for both branches.

Reply:
Oh, interesting I see why. We take the MidiInputMapping as const ref, so in any way to execute a function on the QJSValue, it needs to copy it. I think the constness goes quite deep here, so it probably doesn't make much sense to refactor it to avoid the copy. It's probably quite cheap anyways (at least that's what Qt usually assumes). In that case, you can just take the `const ConfigKey&` in one lambda and by-value (creating a copy in the process) with `QJSValue`. So `[this, pEngine, channel, control, value, status](QJSValue function) {`

Reply:
How do I disable the clang-tidy warning?

Reply:
I'm not sure what you mean, which clang-tidy warning? the clazy warning is fixed by taking the `const ConfigKey&`. The ugly `auto fun = function` is avoided by taking the `QJSValue` by value. You can do both at the same time.

Reply:
Well, clang-tidy complains if I write:

```c++
[this, pEngine, channel, control, value, status](
        QJSValue function) {
    const auto args = QJSValueList{ /* snip **/  };
    if (!pEngine->executeFunction(&function, args)) {  /* snip **/    }
    return;
}}
```

and I get a compiler error if I write:

```c++
[this, pEngine, channel, control, value, status](
        &QJSValue function) {
    const auto args = QJSValueList{ /* snip **/  };
    if (!pEngine->executeFunction(function, args)) {  /* snip **/    }
    return;
}}
```

The only thing that pleases both is:

```c++
[this, pEngine, channel, control, value, status](
        const &QJSValue function) {
    const auto args = QJSValueList{ /* snip **/  };
    auto fun = function;
    if (!pEngine->executeFunction(function, args)) {  /* snip **/    }
    return;
}}
```

which is… meh.

Reply:
I don't understant why `bool ControllerScriptEngineBase::executeFunction` was modified to pass `QJSValue* pFunctionObject` and not `const QJSValue& pFunctionObject`. I thought the second was safer than the first? :thinking:

Reply:
> (Parodn me If I ask too many questions. C++ is still very obscure to me)

No worries. 

> I don't understant why bool ControllerScriptEngineBase::executeFunction was modified to pass QJSValue* pFunctionObject and not const QJSValue& pFunctionObject

Because `ControllerScriptEngineBase::executeFunction` will eventually call `QJSValue::call`, which is non-const. So `executeFunction` can't take a const reference. Otherwise it would have to copy the QJSValue every time before being able to call `call` on it.

> Well, clang-tidy complains if I write:

What clang-tidy error exactly? I don't get any error locally.

Reply:
Hmm… You were right. clang-tidy doesn't complain.